src/hotspot/share/utilities/lockFreeQueue.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_LOCKFREEQUEUE_HPP
+#define SHARE_UTILITIES_LOCKFREEQUEUE_HPP
+
+#include "memory/padded.hpp"
+#include "runtime/atomic.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+// The LockFreeQueue template provides a lock-free FIFO. Its structure
+// and usage is similar to LockFreeStack. It has inner paddings, and
+// optionally use GlobalCounter critical section in pop() to address
+// the ABA problem. This class has a restriction that pop() may return
+// NULL when there are objects in the queue if there is a concurrent
+// push/append operation.
+//
+// \tparam T is the class of the elements in the queue.
+//
+// \tparam next_ptr is a function pointer.  Applying this function to
+// an object of type T must return a pointer to the list entry member
+// of the object associated with the LockFreeQueue type.
+//
+// \tparam rcu_pop true if use GlobalCounter critical section in pop().
+template<typename T, T* volatile* (*next_ptr)(T&), bool rcu_pop>
+class LockFreeQueue {
+  NONCOPYABLE(LockFreeQueue);
+
+protected:
```

Comment:
Protected members (to be accessible from a derived class) are inconsistent with a public non-virtual destructor (that may allow destructor slicing). I dislike classes that try to be both concrete implementation classes and base classes; they are hard to design well (and this class wasn't intended to be such).  This was done to allow G1DirtyCardQueueSet to extend it with the `take_all` function; that seems like a useful operation in the generic form, even if it can't be made thread-safe.  (The G1 function asserts_at_safepoint(), but that's not really appropriate for a generic form.)

Reply:
To clarify, do we want to move take_all() to LockFreeQueue, remove assert_at_safepoint() and put a big warning that it is not thread safe? It perhaps also involves adding the "struct HeadTail" to this class.

Another option is to provide two getter methods that returns T** for &_head and &_tail. They are not thread-safe either, but maybe more generic than take_all().

Reply:
Either of those options seems okay, though the two getters exposes more of the implementation details.  The existing take_all is already described as not thread-safe.  HeadTail => `Pair<T*, T*>` would work.